controller/sharding/consistent/consistent.go

```diff
@@ -0,0 +1,273 @@
+func (c *Consistent) GetLeast(client string) (string, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	if c.clients.Len() == 0 {
+		return "", ErrNoHosts
+	}
+	h := c.hash(client)
+	for {
+		var foundItem btree.Item
+		c.clients.AscendGreaterOrEqual(item{h}, func(bItem btree.Item) bool {
+			if h != bItem.(item).value {
+				foundItem = bItem
+				return false // stop the iteration
+			}
+			return true
+		})
+
+		if foundItem == nil {
+			// If no host found, wrap around to the first one.
+			foundItem = c.clients.Min()
+		}
+		key := c.clients.Get(foundItem)
+		if key != nil {
+			host := c.servers[key.(item).value]
+			if c.loadOK(host) {
+				return host, nil
+			}
+			h = key.(item).value
+		} else {
+			return client, nil
+		}
+	}
+}
+
+// Sets the load of `server` to the given `load`
+func (c *Consistent) UpdateLoad(server string, load int64) {
+	c.Lock()
+	defer c.Unlock()
+
+	if _, ok := c.loadMap[server]; !ok {
+		return
+	}
+	c.totalLoad -= c.loadMap[server].Load
+	c.loadMap[server].Load = load
+	c.totalLoad += load
+}
+
+// Increments the load of host by 1
+//
+// should only be used with if you obtained a host with GetLeast
+func (c *Consistent) Inc(server string) {
+	c.Lock()
+	defer c.Unlock()
+
+	if _, ok := c.loadMap[server]; !ok {
+		return
+	}
+	atomic.AddInt64(&c.loadMap[server].Load, 1)
+	atomic.AddInt64(&c.totalLoad, 1)
+}
+
+// Decrements the load of host by 1
+//
+// should only be used with if you obtained a host with GetLeast
+func (c *Consistent) Done(server string) {
+	c.Lock()
+	defer c.Unlock()
+
+	if _, ok := c.loadMap[server]; !ok {
+		return
+	}
+	atomic.AddInt64(&c.loadMap[server].Load, -1)
+	atomic.AddInt64(&c.totalLoad, -1)
+}
+
+// Deletes host from the ring
+func (c *Consistent) Remove(server string) bool {
+	c.Lock()
+	defer c.Unlock()
+
+	for i := 0; i < c.replicationFactor; i++ {
+		h := c.hash(fmt.Sprintf("%s%d", server, i))
+		delete(c.servers, h)
+		c.delSlice(h)
+	}
+	delete(c.loadMap, server)
+	return true
+}
+
+// Return the list of servers in the ring
+func (c *Consistent) Servers() (servers []string) {
+	c.RLock()
+	defer c.RUnlock()
+	for k := range c.loadMap {
+		servers = append(servers, k)
+	}
+	return servers
+}
+
+// Returns the loads of all the hosts
+func (c *Consistent) GetLoads() map[string]int64 {
+	loads := map[string]int64{}
+
+	for k, v := range c.loadMap {
+		loads[k] = v.Load
+	}
+	return loads
+}
+
+// Returns the maximum load of the single host
+// which is:
+// (total_load/number_of_hosts)*1.25
+// total_load = is the total number of active requests served by hosts
+// for more info:
+// https://research.googleblog.com/2017/04/consistent-hashing-with-bounded-loads.html
+func (c *Consistent) MaxLoad() int64 {
+	if c.totalLoad == 0 {
+		c.totalLoad = 1
+	}
+	var avgLoadPerNode float64
+	avgLoadPerNode = float64(c.totalLoad / int64(len(c.loadMap)))
+	if avgLoadPerNode == 0 {
+		avgLoadPerNode = 1
+	}
+	avgLoadPerNode = math.Ceil(avgLoadPerNode * 1.25)
```

Comment:
Is `1.25` always going to be constant or can this be configurable? Never the less, should we move this value to a constant?

Reply:
For now I would move this to a constant and not make it configurable. 1.25 is the load factor described by the original paper that the algorithm is based on and was found to be the ideal balance between keeping the shards uniform while also keeping consistency when changing shard numbers.

Reply:
Thanks @andklee for the explanation. I agree to moving it to a constant and having a comment to describe what it represents.

Reply:
Extracted as a constant and documenting the constant value